libraries/chain/include/eosio/chain/hotstuff/finalizer.hpp

```diff
@@ -0,0 +1,149 @@
+#include <eosio/chain/hotstuff/hotstuff.hpp>
+#include <eosio/chain/hotstuff/finalizer_policy.hpp>
+#include <eosio/chain/fork_database.hpp>
+#include <fc/crypto/bls_utils.hpp>
+#include <fc/io/cfile.hpp>
+#include <fc/reflect/reflect.hpp>
+#include <fc/crypto/sha256.hpp>
+#include <compare>
+#include <utility>
+
+namespace eosio::chain {
+   // ----------------------------------------------------------------------------------------
+   struct qc_chain_t {
+      block_state_ptr b2; // first phase,  prepare
+      block_state_ptr b1; // second phase, precommit
+      block_state_ptr b;  // third phase,  commit
+   };
+
+   // ----------------------------------------------------------------------------------------
+   struct finalizer {
+      enum class vote_decision { strong_vote, weak_vote, no_vote };
+
+      struct proposal_ref {
+         block_id_type         id;
+         block_timestamp_type  timestamp;
+
+         proposal_ref() = default;
+
+         template<class BSP>
+         proposal_ref(const BSP& p) :
+            id(p->id()),
+            timestamp(p->timestamp())
+         {}
+
+         proposal_ref(const block_id_type& id, block_timestamp_type t) :
+            id(id), timestamp(t)
+         {}
+
+         void reset() {
+            id = block_id_type();
+            timestamp = block_timestamp_type();
+         }
+
+         bool empty() const { return id.empty(); }
+
+         explicit operator bool() const { return !id.empty(); }
+
+         auto operator==(const proposal_ref& o) const {
+            return id == o.id && timestamp == o.timestamp;
+         }
+      };
+
+      struct safety_information {
+         block_timestamp_type last_vote_range_start;
+         proposal_ref         last_vote;          // v_height under hotstuff
+         proposal_ref         lock;               // b_lock under hotstuff
+
+         static constexpr uint64_t magic = 0x5AFE11115AFE1111ull;
+
+         static safety_information unset_fsi() { return {block_timestamp_type(), {}, {}}; }
+
+         auto operator==(const safety_information& o) const {
+            return last_vote_range_start == o.last_vote_range_start &&
+               last_vote == o.last_vote &&
+               lock == o.lock;
+         }
+      };
+
+      bls_private_key           priv_key;
+      safety_information        fsi;
+
+   private:
+      using branch_type = fork_database_if_t::branch_type;
+      qc_chain_t     get_qc_chain(const block_state_ptr& proposal, const branch_type& branch) const;
+      vote_decision  decide_vote(const block_state_ptr& proposal, const fork_database_if_t& fork_db);
+
+   public:
+      std::optional<vote_message> maybe_vote(const bls_public_key& pub_key, const block_state_ptr& bsp,
+                                             const digest_type& digest, const fork_database_if_t& fork_db);
+   };
+
+   // ----------------------------------------------------------------------------------------
+   struct my_finalizers_t {
```

Comment:
I thought this name again. It seems `my_fianlizers` is more accurate, as we are talking about the finalizers on the local node. In configuration instructions to BPs, local finalizer sounds more natural than my finalizers. `my` implies something at implementation level like a class or a file. @heifner @arhag ?

Reply:
This is a class name at the implementation level.